driver/src/gy521.cpp

```diff
@@ -0,0 +1,91 @@
+/**
+ * @file include/gy521.cpp
+ * @brief Drivers for the GY-521 6 axis accelerometer.
+ */
+
+#include "gy521.hpp"
+
+#include <chrono>
+#include <cstdint>
+
+#include "defines.hpp"
+
+static constexpr uint8_t REG_INTERRUPT_STATUS = 0x3A;
+static constexpr uint8_t REG_ACCEL_X = 0x3B;
+static constexpr uint8_t REG_ACCEL_Y = 0x3D;
+static constexpr uint8_t REG_ACCEL_Z = 0x3F;
+static constexpr uint8_t REG_GYRO_X = 0x43;
+static constexpr uint8_t REG_GYRO_Y = 0x45;
+static constexpr uint8_t REG_GYRO_Z = 0x47;
+static constexpr uint8_t REG_PWR_MGMT_1 = 0x6B;
+static constexpr uint8_t REG_WHOAMI = 0x75;
+
+gy521::gy521(
+	const gpiod::chip &chip, uint32_t int_pin, uint8_t dev_addr, int adapter)
+	: interrupt(chip.get_line(int_pin)),
+	  i2cd(i2c(dev_addr, adapter)) {
+	interrupt.request({
+		.consumer = GPIO_CONSUMER,
+		.request_type = gpiod::line_request::EVENT_RISING_EDGE,
+		.flags = 0,
+	});
+
+	uint8_t whoami = i2cd.read_byte(REG_WHOAMI);
+	// 0x68 is the expected return value
+	if (whoami != 0x68) {
+		printf("I2C ERROR, WHOAMI RESULT: %X", whoami);
+	}
+
+	uint8_t wakeup = 0x00;
+	if (!i2cd.write(REG_PWR_MGMT_1, &wakeup, 1)) {
+		printf("WAKEUP WRITE FAIL");
+	}
+}
+
+gy521::~gy521() {
+	// sleep device
+	uint8_t sleep = 0x40;
+	i2cd.write(REG_PWR_MGMT_1, &sleep, 1);
+	active = false;
+	int_thread->join();
+	interrupt.release();
+}
+
+void gy521::on_interrupt(std::function<void()> callback) {
```

Comment:
Calling this function twice would be bad
Need to kill existing thread before making a new one